include/ghex/unstructured/pattern.hpp

```diff
@@ -191,129 +175,435 @@ namespace gridtools {
+                                            is.push_back(all_recv_indices[static_cast<std::size_t>(other_rank)][recv_indices_idx] -
+                                                    rank_local_start_index +
+                                                    d.inner_size()); // index offset
+                                        }
+                                    }
+                                    if (is.size()) {
+                                        index_container_type ic{is};
+                                        my_patterns[p].recv_halos().insert(std::make_pair(id, ic));
+                                    }
+                                    recv_indices_start_idx +=
+                                            static_cast<std::size_t>(all_recv_counts[static_cast<std::size_t>(other_rank)][static_cast<std::size_t>(other_domain_idx)]);
+                                }
+                            }
                         }
 
-                        // set up send halos
-                        comm.all_to_allv(recv_indexes, recv_counts, recv_displs,
-                                send_indexes, send_counts, send_displs);
-                        for (std::size_t rank = 0; rank < size; ++rank) {
-                            if (send_counts[rank]) {
-                                // WARN: very simplified definition of extended domain id;
-                                // a more complex one is needed for multiple domains
-                                int tag = (my_address << 7) + rank; // WARN: maximum rank / address = 2^7 - 1
-                                extended_domain_id_type id{static_cast<int>(rank), static_cast<int>(rank), static_cast<address_type>(rank), tag};
-                                std::vector<index_type, u_m_allocator_t> remote_index{};
-                                remote_index.resize(send_counts[rank]);
-                                std::memcpy(&remote_index[0],
-                                        &send_indexes[static_cast<std::size_t>(send_displs[rank])],
-                                        static_cast<std::size_t>(send_counts[rank]) * sizeof(index_type));
-                                index_container_type ic{ {static_cast<int>(rank), std::move(remote_index), send_levels[rank]} };
-                                p.send_halos().insert(std::make_pair(id, ic));
-                            }
+                        rank_local_start_index += halo_size;
+                    }
+
+                    return pattern_container<communicator_type, grid_type, domain_id_type>(std::move(my_patterns), m_max_tag);
+
+                }
+
+                template<typename Transport, typename ThreadPrimitives, typename HaloGenerator, typename RecvDomainIdsGen, typename DomainRange>
+                static auto apply(tl::context<Transport, ThreadPrimitives>& context, HaloGenerator&& hgen, RecvDomainIdsGen&& recv_domain_ids_gen, DomainRange&& d_range) {
```

Comment:
These functions are huge. Is it possible to write somewhere a description to understand what's going on in an intuitive way?

Reply:
Added some documentation for these two functions. Can go more in detail if needed.